src/vdaf/hits.rs

```diff
@@ -0,0 +1,741 @@
+use crate::fp::log2;
+use crate::prng::Prng;
+use crate::vdaf::suite::{Key, KeyDeriver, KeyStream, Suite};
+use crate::vdaf::{Share, VdafError};
+
+/// An input for an IDPF ([`Idpf`]).
+#[derive(Clone, Debug, Hash, Eq)]
+pub struct IdpfInput<'a> {
+    data: &'a [u8],
+    len: usize,
+}
+
+impl<'a> IdpfInput<'a> {
+    /// Constructs an IDPF input using the first `len` bits of `data`.
+    pub fn new(data: &'a [u8], len: usize) -> Result<Self, VdafError> {
+        if len > data.len() << 3 {
+            return Err(VdafError::Uncategorized(format!(
+                "desired bit length ({} bits) exceeds data length ({} bytes)",
+                len,
+                data.len()
+            )));
+        }
+
+        Ok(Self { data, len })
+    }
+
+    /// Return the `bit`-th bit of the input. Bounds checking is performed by the caller.
+    fn bit(&self, bit: usize) -> usize {
+        let i = bit >> 3;
+        let j = ((i << 3) ^ bit) as u8;
+        (self.data[i] >> j) as usize & 1
+    }
+
+    /// Return the index of input `self` in the look-up table of a `ToyIdpf` instance.
+    fn index(&self) -> usize {
+        let mut index = 1 << self.len;
+        for i in 0..self.len {
+            index += self.bit(i) << i;
+        }
+        index - 1
+    }
+
+    /// Construct a new input that is a prefix of `self`. Bounds checking is performed by the
+    /// caller.
+    fn prefix(&self, len: usize) -> Self {
+        Self {
+            data: self.data,
+            len,
+        }
+    }
+}
+
+impl<'a> PartialEq<IdpfInput<'a>> for IdpfInput<'a> {
+    fn eq(&self, other: &IdpfInput) -> bool {
+        if self.len != other.len {
+            return false;
+        }
+
+        // Do a constant-time comparison, masking the last byte to fit the input length (in bits).
+        let q = self.len / 8;
+        let m = (self.len % 8) as u8;
+        let mut out = 0;
+        for (i, (a, b)) in self.data.iter().zip(other.data).enumerate() {
+            let (x, y) = match i == q {
+                true => (*a & m, *b & m),
+                false => (*a, *b),
+            };
+            out |= x ^ y;
+        }
+        out == 0
+    }
+}
+
+/// An Incremental Distributed Point Function (IDPF), as defined by [[BBCG+21]].
+///
+/// [BBCG+21]: https://eprint.iacr.org/2021/017
+//
+// NOTE(cjpatton) The real IDPF API probably needs to be stateful.
+pub trait Idpf<const KEY_LEN: usize, const OUT_LEN: usize>: Sized {
+    /// The finite field over which the IDPF is defined.
+    //
+    // NOTE(cjpatton) The IDPF of [BBCG+21] might use different fields for different levels of the
+    // prefix tree.
+    type Field: FieldElement;
+
+    /// Generate a new sequence of IDPF shares for `input`. Parameter `output` is an iterator that
+    /// is invoked to get the output value for each successive level of the prefix tree.
+    fn gen<O: IntoIterator<Item = Self::Field>>(
+        input: &IdpfInput,
+        output: O,
```

Comment:
It's a bit confusing having an input parameter named `output` and then some other value returned from the function. I need to read the HH paper to better understand how all this fits together, but at least a doccomment explaining what value is returned from this function would help.

Reply:
No, I think it's reasonable to expect readers or editors of this code to have read the paper and be familiar with IDPFs.